feat(Model): Add model through issue forms

gaiborjosue opened the pull request:
Hello,

This pull request will close #76 #80 #81 

This enhancement or new feature uses GitHub issue forms to enable users to submit a new model more easily and automatically.

This means that the new workflow will provide enhancement for:
1. Organization. We are responsible for the repository structure since we ask the users for specific information. The user only has to worry about providing URLs
2. Standardize. Since we will test the main Python script, we ensure all models being added follow a certain guideline or distribution of flags.
3. Testing. This workflow tests Python scripts with EC2, singularity, and docker. This means we ensure their Python scripts work before adding them to the zoo.
4. User friendly. The workflow counts with a notification/tag system that will let the user know if their request failed or had success. Also, it will automatically create a new branch for the issue and link a draft PR. So that when it is ready and has passed the checks, users can open the PR. 

Next, I will attach a flow chart describing a synthesized version of the workflow since it has +500 lines:
![flow_addmodel](https://github.com/neuronets/trained-models/assets/78832141/eeba6aa8-a3f9-43f2-bb17-db5a8a0c3886)

As described above, the user will create an issue form. The action will run. If it fails, it will tag the issue as failed, and the user will have to fix issues and then change the label to "Ready-to-test" for the workflow to trigger and test again.

During these processes, the workflow also has metric tracking and useful performance insights that users may check during the workflow execution in the specific Job Summary. 

Check how the templates look here: https://github.com/gaiborjosue/trained-models-fork/issues/new/choose

@satra This new feature is 100% tested with a testing model, 'DeepCSR.'

However, a few steps are missing before this PR is/should be merged. We are working on these and will finish them as fast as possible. Perhaps @hvgazula can provide more insights into some of the following steps:

- [ ] As per @hvgazula suggestion, we will use composite actions to reduce the line quantity of the workflow. Even though it is stratified within multiple jobs and has comments, it may still be hard to read.
- [ ] Due to some issues mentioned here #82 and further concerns @hvgazula may address in this PR, such as CUDA versions, we also want to test it with PialNN. Since PialNN does not use a config file, we want to test scenarios for both models: with the config file and without a config file.
- [ ] This workflow goes hand-in-hand with another issue form, the Update Model. However, I only have to sync functionalities/steps added during this week from the Add Model to the Update Model workflow. I will do this as soon as possible.
- [ ]  We have to add secrets and variables from EC2 so that the workflow can access a machine and work properly.
- [ ] Implement docker image building in the EC2 machine as well. This will improve performance (in time) because GitHub action's machine takes around 1 hour to build a docker image.
- [ ] Clean up old workflows, such as doing it through a PR.
- [ ] Update the readme file with a specific explanation of using this feature and other relevant recommendations/requirements.
- [ ] Add tags: "failed," "success," and "Ready-to-test" to the issues menu so the workflow can tag issues correctly.

As mentioned, the "Add a new model" feature is 100% ready :). There are just a few steps before this PR can be merged.

hvgazula commented:
Impressive. Thank you very much for all your efforts @gaiborjosue 👍. Now, let's wait for @satra's feedback.

satra commented:
i'm fine with implementing these. however, i would urge to create an MVP that adds models even if they fail for various compatibility issues. i.e. get models that work in and tell users about models that don't work. that's all that is needed. leave compatibility to the user, just tell them what environment things are running in.

hvgazula commented:
I have two models that are ready to be pushed to the repo. Will give it a try next week. @gaiborjosue Let's schedule a time next week to do one.

gaiborjosue commented:
@hvgazula of course👍

hvgazula reviewed (commented):
Some generic comments which I made last week but forgot to push.

hvgazula commented:
@gaiborjosue If you don't mind, can you be a little more descriptive with your commit messages (if possible)?

gaiborjosue commented:
Hello @hvgazula, yes. Sorry about that. Since some of them are very small changes and they require a push to sync with the workflows, I do small messages. I will dedicate a few more seconds to better commits. Thank you for the feedback.

gaiborjosue commented:
Hello @hvgazula, the docs are updated, and workflows are tested.

General changes since the last update:

1. Added two new workflows to assign the user to the repo (since I realized the user can't auto-assign) and auto-tag the issue when commented "Ready-to-test" (since the user can't assign label).
2. Allowed multiple weights/sample datasets to be uploaded via zip files.
3. Tested PialNN successfully.

Thanks,
Edward

hvgazula commented on add_model_instructions.md:
````diff
@@ -21,8 +21,8 @@ Carefully fill out the issue template with the required information:
-- Paste a direct download link from Google Drive to the model's weights.
-  Example URL structure: ```"https://drive.google.com/uc?id=<FILEID>&confirm=t"```
+- Paste a direct download link to the model's weights (Google Drive, Github Raw, etc.). If you need to upload multiple weights, zip the weights and paste the download link here.
+  Example URL structure for Google Drive: ```"https://drive.google.com/uc?id=<FILEID>&confirm=t"```
````
@gaiborjosue  You might want to provide details about extracting FILEID.

gaiborjosue commented:
Hello @hvgazula, I added comments on file id extraction 👍

hvgazula reviewed (approved):
LGTM

gaiborjosue commented:
Hello @hvgazula, nice. Now we need to add the env variables and secrets. This should be related to Ec2 and a github token with issue and workflow permissions.

hvgazula commented on .github/workflows/new_model.yml:
```diff
@@ -0,0 +1,629 @@
+        run: |
+          # Parent directory as a bind path in a env variable
+
+          singularity exec --nv --bind /actions-runner,/actions-runner/_work/,/actions-runner/_work/${{ github.event.repository.name }}/${{ github.event.repository.name }}/${{ needs.create_new_branch.outputs.MODELPATH }}/,/actions-runner/_work/${{ github.event.repository.name }}/${{ github.event.repository.name }}/${{ needs.create_new_branch.outputs.MODELPATH }}/docker,/actions-runner/_work/${{ g ...
```
@gaiborjosue Here's a comment I made in the past that went unresolved. Probably skipped your attention.
